opal/corelib/kernel.rb

```diff
@@ -233,6 +248,88 @@ def extend(*mods)
     self
   end
 
+  def freeze
+    return self if frozen?
+
+    %x{
+      self.$$frozen = true;
+
+      if (typeof(self) === "object") {
+        // make setters of instance variables throw FrozenError
+        var prop, prop_type, current_value, desc;
+
+        // Optimization for Array because iterating through props below
+        // would iterate through array members. Consequently instance
+        // variables in Arrays stay mutable when accessed directly
+        // even when the array is frozen.
+        if (!self.$$is_array) {
+          for(prop in self) {
+            prop_type = typeof(prop);
+
+            // prop_type "object" here is a String(), skip $ props
+            if ((prop_type === "string" || prop_type === "object") && prop.startsWith('$')) {
+              continue;
+            }
+
+            desc = Object.getOwnPropertyDescriptor(self, prop);
+            if (desc && desc.enumerable && desc.writable) {
+              current_value = self[prop];
+
+              // create closure for s and v
+              // for Opal 2.0 let for s and v should do the trick, instead of a function
+              (function() {
+                // set v to undefined, as if the property is not set
+                var s = false, p = prop, v = undefined;
```

Comment:
This logic with single letter variables is as we understand for debugging... `s` is always false, right?

Reply:
Oh, i see, i think i can simplify this a bit more ... i will try ...

Reply:
p was left over, s is now not required at all